cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/LocalizationUtils.java

```diff
@@ -0,0 +1,152 @@
+   * @return the {@link File} pointing to the localized file.
+   */
+  public static File localizeResource(String fileName, LocalizeResource resource, File targetDir) throws IOException {
+    File tempFile = new File(targetDir, fileName);
+    File fileToLocalize = new File(resource.getURI().getPath());
+    if (resource.isArchive()) {
+      LOG.debug("Uncompress file {} to {}", fileToLocalize, tempFile);
+      uncompress(fileToLocalize, tempFile);
+    } else {
+      LOG.debug("Copy file from {} to {}", fileToLocalize, tempFile);
+      Files.copy(fileToLocalize, tempFile);
+    }
+    return tempFile;
+  }
+
+  private static void uncompress(File archive, File targetDir) throws IOException {
+    if (!targetDir.exists()) {
+      //noinspection ResultOfMethodCallIgnored
+      targetDir.mkdir();
+    }
+    String extension = Files.getFileExtension(archive.getPath());
+    switch (extension) {
+      case "zip":
+        unzip(archive, targetDir);
+        break;
+      case "jar":
+        unzip(archive, targetDir);
+        break;
+      case "gz":
+        // gz is not recommended for archiving multiple files together. So we only support .tar.gz
+        untargz(archive, targetDir);
+        break;
+      case "tar":
+        untar(archive, targetDir);
+        break;
+      default:
+        throw new IllegalArgumentException(String.format("Unsupported compression type %s. Only 'zip', 'jar', " +
+                                                           "'gz', 'gzip', 'tar', 'tar.gz' are supported.",
+                                                         extension));
+    }
+  }
+
+  private static void unzip(File zipFile, File targetDir) throws IOException {
+    final ZipFile zip = new ZipFile(zipFile);
+    Enumeration<? extends ZipEntry> zipEntries = zip.entries();
+    while (zipEntries.hasMoreElements()) {
+      final ZipEntry entry = zipEntries.nextElement();
+      // Have to do a new File(file.getName()).getName() to get the file name, because for ZipFile, entry.getName
+      // returns the absolute path.
+      String outFileName = new File(entry.getName()).getName();
+      File outFile = new File(targetDir, outFileName);
+      if (entry.isDirectory()) {
+        //noinspection ResultOfMethodCallIgnored
+        outFile.mkdir();
+        continue;
+      }
+      Files.copy(new InputSupplier<InputStream>() {
+        @Override
+        public InputStream getInput() throws IOException {
+          return zip.getInputStream(entry);
+        }
+      }, outFile);
+    }
+  }
+
+  private static void untar(File tarFile, File targetDir) throws IOException {
+    try (FileInputStream fis = new FileInputStream(tarFile);
+         TarArchiveInputStream tis = new TarArchiveInputStream(fis)) {
+      extractTar(tis, targetDir);
+    }
+  }
+
+  private static void untargz(File tarGzFile, File targetDir) throws IOException {
+    try (FileInputStream fis = new FileInputStream(tarGzFile);
```

Comment:
You don't need `fis` and `gzis`. Just do `TarArchiveInputStream is = new TarArchiveInputStream(new GZIPInputStream(new FileInputStream(file)))`